File: data_structures/kotlin/heaps.kt
import java.util.*

class Heap<T>(private val isMinHeap: Boolean = false, private val comparator: Comparator<T>) {
    private val elements: MutableList<T> = mutableListOf()

    // MARK: - Insertion

    fun insert(element: T) {
        elements.add(element)
        heapifyUp()
    }

    private fun heapifyUp() {
        var currentIndex = elements.size - 1
        while (currentIndex > 0) {
            val parentIndex = (currentIndex - 1) / 2

            if (shouldSwap(parentIndex, currentIndex)) {
                Collections.swap(elements, parentIndex, currentIndex)
                currentIndex = parentIndex
            } else {
                break
            }
        }
    }

    // MARK: - Removal

    fun remove(): T? {
        if (elements.isEmpty()) {
            return null
        }

        if (elements.size == 1) {
            return elements.removeAt(0)
        }

        val root = elements[0]
        elements[0] = elements.removeAt(elements.size - 1)
        heapifyDown()

        return root
    }

    private fun heapifyDown() {
        var currentIndex = 0

        while (hasLeftChild(currentIndex)) {
            var smallestChildIndex = leftChildIndex(currentIndex)

            if (hasRightChild(currentIndex) && shouldSwap(rightChildIndex(currentIndex), smallestChildIndex)) {
                smallestChildIndex = rightChildIndex(currentIndex)
            }

            if (shouldSwap(smallestChildIndex, currentIndex)) {
                Collections.swap(elements, smallestChildIndex, currentIndex)
                currentIndex = smallestChildIndex
            } else {
                break
            }
        }
    }

    // MARK: - Peek

    fun peek(): T? {
        return elements.firstOrNull()
    }

    // MARK: - Helper Methods

    private fun leftChildIndex(parentIndex: Int): Int {
        return 2 * parentIndex + 1
    }

    private fun rightChildIndex(parentIndex: Int): Int {
        return 2 * parentIndex + 2
    }

    private fun hasLeftChild(index: Int): Boolean {
        return leftChildIndex(index) < elements.size
    }

    private fun hasRightChild(index: Int): Boolean {
        return rightChildIndex(index) < elements.size
    }

    private fun shouldSwap(firstIndex: Int, secondIndex: Int): Boolean {
        return if (isMinHeap) {
            comparator.compare(elements[firstIndex], elements[secondIndex]) > 0
        } else {
            comparator.compare(elements[firstIndex], elements[secondIndex]) < 0
        }
    }

    // MARK: - Build Heap

    fun buildHeap(elements: List<T>) {
        this.elements.clear()
        this.elements.addAll(elements)
        for (i in (elements.size / 2) - 1 downTo 0) {
            heapifyDown(i)
        }
    }

    // MARK: - Replace Root

    fun replaceRoot(element: T) {
        if (elements.isEmpty()) {
            insert(element)
            return
        }

        elements[0] = element
        heapifyDown(0)
    }

    // MARK: - Remove at Index

    fun removeAtIndex(index: Int): T? {
        if (index < 0 || index >= elements.size) {
            return null
        }

        if (index == elements.size - 1) {
            return elements.removeAt(elements.size - 1)
        }

        Collections.swap(elements, index, elements.size - 1)
        val removedElement = elements.removeAt(elements.size - 1)
        heapifyDown(index)

        return removedElement
    }

    // MARK: - Sort

    fun sort(): List<T> {
        val sortedElements = mutableListOf<T>()
        while (elements.isNotEmpty()) {
            sortedElements.add(remove()!!)
        }
        return sortedElements
    }

    // MARK: - Index for Element

    fun index(element: T): Int? {
        return index(element, 0)
    }

    private fun index(element: T, startIndex: Int): Int? {
        if (startIndex < elements.size) {
            if (elements[startIndex] == element) {
                return startIndex
            }

            leftChildIndex(startIndex).let {
                if (it < elements.size) {
                    val result = index(element, it)
                    if (result != null) return result
                }
            }

            rightChildIndex(startIndex).let {
                if (it < elements.size) {
                    val result = index(element, it)
                    if (result != null) return result
                }
            }
        }

        return null
    }

    // ... Add more heap operations as needed ...
}

fun main() {
    // Example usage
    val maxHeap = Heap(isMinHeap = false, comparator = compareBy { it })
    maxHeap.insert(3)
    maxHeap.insert(1)
    maxHeap.insert(4)
    maxHeap.insert(1)
    maxHeap.insert(5)
    maxHeap.insert(9)

    println("Max Heap: ${maxHeap.elements}")

    val minHeap = Heap(isMinHeap = true, comparator = compareBy { it })
    minHeap.insert(3)
    minHeap.insert(1)
    minHeap.insert(4)
    minHeap.insert(1)
    minHeap.insert(5)
    minHeap.insert(9)

    println("Min Heap: ${minHeap.elements}")
}

File: data_structures/python/heaps.py
import heapq

class Heap:
    def __init__(self, is_min_heap=False):
        self.elements = []
        self.is_min_heap = is_min_heap
        self.comparator = (lambda a, b: a < b) if is_min_heap else (lambda a, b: a > b)

    # MARK: - Insertion

    def insert(self, element):
        heapq.heappush(self.elements, element)
        self._heapify_up()

    def _heapify_up(self):
        index = len(self.elements) - 1
        while index > 0:
            parent_index = (index - 1) // 2

            if self._should_swap(parent_index, index):
                self.elements[parent_index], self.elements[index] = self.elements[index], self.elements[parent_index]
                index = parent_index
            else:
                break

    # MARK: - Removal

    def remove(self):
        if self.elements:
            root = self.elements[0]
            last_element = self.elements.pop()
            if self.elements:
                self.elements[0] = last_element
                self._heapify_down()
            return root
        return None

    def _heapify_down(self):
        index = 0
        while self._has_left_child(index):
            smallest_child_index = self._left_child_index(index)

            if self._has_right_child(index) and self._should_swap(smallest_child_index, self._right_child_index(index)):
                smallest_child_index = self._right_child_index(index)

            if self._should_swap(index, smallest_child_index):
                self.elements[index], self.elements[smallest_child_index] = self.elements[smallest_child_index], self.elements[index]
                index = smallest_child_index
            else:
                break

    # MARK: - Peek

    def peek(self):
        return self.elements[0] if self.elements else None

    # MARK: - Helper Methods

    def _left_child_index(self, index):
        return 2 * index + 1

    def _right_child_index(self, index):
        return 2 * index + 2

    def _has_left_child(self, index):
        return self._left_child_index(index) < len(self.elements)

    def _has_right_child(self, index):
        return self._right_child_index(index) < len(self.elements)

    def _should_swap(self, index1, index2):
        return self.comparator(self.elements[index1], self.elements[index2])

    # MARK: - Build Heap

    def build_heap(self, elements):
        self.elements = elements[:]
        heapq.heapify(self.elements)

    # MARK: - Replace Root

    def replace_root(self, element):
        if self.elements:
            heapq.heappushpop(self.elements, element)
        else:
            self.insert(element)

    # MARK: - Remove at Index

    def remove_at_index(self, index):
        if 0 <= index < len(self.elements):
            element = self.elements[index]
            self.elements[index] = self.elements[-1]
            self.elements.pop()
            self._heapify_down(index)
            return element
        return None

    # MARK: - Sort

    def sort(self):
        sorted_elements = []
        while self.elements:
            sorted_elements.append(self.remove())
        return sorted_elements

    # MARK: - Index for Element

    def index(self, element):
        try:
            return self.elements.index(element)
        except ValueError:
            return None

    # ... Add more heap operations as needed ...

# Example usage
max_heap = Heap()
max_heap.insert(3)
max_heap.insert(1)
max_heap.insert(4)
max_heap.insert(1)
max_heap.insert(5)
max_heap.insert(9)

print("Max Heap:", max_heap.elements)

min_heap = Heap(is_min_heap=True)
min_heap.insert(3)
min_heap.insert(1)
min_heap.insert(4)
min_heap.insert(1)
min_heap.insert(5)
min_heap.insert(9)

print("Min Heap:", min_heap.elements)
